ui/src/source.ts

```diff
@@ -379,29 +297,71 @@ export const revisionQueryEq = (
   }
 };
 
-export const readme = (
-  projectId: string,
-  peerId: string,
-  revision: RevisionQuery
-): Readable<remote.Data<Readme | null>> => {
-  const readme = remote.createStore<Readme | null>();
-
-  remote
-    .chain(objectStore.readable, readme)
-    .then((object: SourceObject) => {
-      if (object.info.objectType === ObjectType.Tree) {
-        const path = findReadme(object as Tree);
-
-        if (path) {
-          return blob(projectId, peerId, revision, path, false);
-        }
-      }
-
-      return null;
-    })
-    .then(blob => (blob && !blob.binary ? blob : null))
-    .then(readme.success)
-    .catch(readme.error);
+export const fetchReadme = (
+  projectUrn: Urn,
+  peerId: PeerId,
+  revision: Revision,
+  tree: Tree
+): Promise<Readme | null> => {
+  const path = findReadme(tree);
+
+  return new Promise((resolve, _reject) => {
```

Comment:
We should avoid explicit `Promise` construction as it is unidiomatic. You can make the function `async` and use the following body.
```typescript
if (!path) {
  return null;
}

let blob;
try {
  blob = await fetchBlob(projectUrn, peerId, path, revision, false);
} catch {
  return null;
}

if (blob && !blob.binary) {
  return blob;
} else {
  return null;
}
```

Reply:
> We should avoid explicit `Promise` construction as it is unidiomatic. You can make the function `async` and use the following body.

Could you expand what is unidiomatic about `Promise` construction? I fail to see how your proposal is an improvement, it introduces an additional variable and a handful more branches, likely harder to reason.

Reply:
I personally find the control flow in my suggestion much clearer. For example we can write `return null` instead of `return resolve(null)`. Using `new Promise` introduces callback-style code for control flow which the whole JS ecosystem has moved away from because it is usually harder to reason about.

> it introduces an additional variable

I don’t think so. `blob` is also introduced in your code as a callback function argument.

> and a handful more branches,

This is not the case either. In your code the branches are just less visible (the ternary operator) and a `.catch` call.